README.md
![forge-cli](https://www.arcblock.io/.netlify/functions/badge/?text=Forge%20CLI)

> Command line toolbox maintained by [Arcblock](https://www.arcblock.io) that helps developers to work with [Forge SDK](https://docs.arcblock.io/en/docs/instruction/sdk)

## Table of Contents

- [Introduction](#introduction)
- [QuickStart](#quickstart)
- [Requirements](#requirements)
- [Install](#install)
- [Usage](#usage)
- [FAQ](#faq)

## Introduction

[![](https://img.shields.io/npm/v/@arcblock/forge-cli.svg?label=forge-cli&style=flat-square)](https://www.npmjs.com/package/@arcblock/forge-cli)
[![docs](https://img.shields.io/badge/powered%20by-arcblock-green.svg?style=flat-square)](https://docs.arcblock.io/en/handbook/)
[![Build Status](https://img.shields.io/travis/com/arcblock/forge-cli?style=flat-square)](https://travis-ci.com/arcblock/forge-cli)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](http://makeapullrequest.com)
[![Gitter](https://img.shields.io/gitter/room/ArcBlock/community?style=flat-square&color=%234cb696)](https://gitter.im/ArcBlock/community?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge)

`forge-cli` is an awesome toolbox for developers to work with forge, and forge is an awesome framework for building decentralized applications. Out of the box forge-cli support following features:

- Manage forge kernel release
- Manage local forge node, join a remote forge powered network
- Create wallets and accounts on blockchain
- Read/subscribe accounts/blocks/transactions/assets on blockchain
- Send transactions to the blockchain
- Use forge components such as forge-web
- Compile and deploy transaction protocol
- Do stake to node/user/asset
- Bootstrap dApps with starters from [here](https://github.com/ArcBlock/forge-dapp-starters)

## Requirements

- Linux/Mac Command Line, **windows is not support currently**, [iTerm](http://www.iterm2.com/) is recommended.
- [Node.js](https://nodejs.org/): please install using [nvm](https://github.com/creationix/nvm), >= v10 && <= v12.x

## QuickStart

[![asciicast](https://asciinema.org/a/280694.svg)](https://asciinema.org/a/280694)

## Install

```shell
npm install -g @arcblock/forge-cli
# OR
yarn global add @arcblock/forge-cli
```

Now `forge` command is available to all new shell sessions.

## Usage

Run `forge` and get available options and subcommands.

```terminal
❯ forge

██████╗ ██╗   ██╗     █████╗ ██████╗  ██████╗██████╗ ██╗      ██████╗  ██████╗██╗  ██╗
██╔══██╗╚██╗ ██╔╝    ██╔══██╗██╔══██╗██╔════╝██╔══██╗██║     ██╔═══██╗██╔════╝██║ ██╔╝
██████╔╝ ╚████╔╝     ███████║██████╔╝██║     ██████╔╝██║     ██║   ██║██║     █████╔╝ 
██╔══██╗  ╚██╔╝      ██╔══██║██╔══██╗██║     ██╔══██╗██║     ██║   ██║██║     ██╔═██╗ 
██████╔╝   ██║       ██║  ██║██║  ██║╚██████╗██████╔╝███████╗╚██████╔╝╚██████╗██║  ██╗
╚═════╝    ╚═╝       ╚═╝  ╚═╝╚═╝  ╚═╝ ╚═════╝╚═════╝ ╚══════╝ ╚═════╝  ╚═════╝╚═╝  ╚═╝
                                                                                      
Usage: forge [options] [command]

Options:
  -V, --version                    output the version number
  -v, --verbose                    Output runtime info when execute subcommand, useful for debug
  -c, --chain-name <chainName>     Execute command use specific chain
  -i, --config-path <path>         Forge config used when starting forge node and initializing gRPC clients
  -r, --npm-registry <registry>    Specify a custom npm registry
  -y, --yes                        Assume that the answer to any confirmation question is yes
  -d, --defaults                   Run command using default values for all questions
  -m, --mirror <url>               Mirror host used to download forge release
  -g, --socket-grpc <endpoint>     Socket gRPC endpoint to connect, with this you can use forge-cli with a remote node
  -h, --help                       output usage information

Commands:
  account <address>                Get an account info by address
  asset <address>                  Get asset info by address
  block [options] [height]         Get the block info from the running node
  blocklet:init [options]          Init a blocklet project
  blocklet:use [options]           Download and install a blocklet
  chain:config [options] [action]  Read/write chain/node config
  chain:create [chainName]         Create a new chain instance
  chain:ls                         List all chains
  chain:remove <chainName>         Remove chain state and config
  chain:reset <chainName>          Reset chain state, but keeps the config
  chain:validator [options]        Update(add, remove, change) or list validators
  config [options] [key] [value]   Config forge cli configs
  declare:node [options]           Declare the current node to be a validator candidate
  deploy:prepare [options]         Prepare node for deploying a multi-node chain
  download [options] [version]     Download a forge release without activate it
  help [subcommand]                Show help of a sub command
  install [options] [version]      Download and setup forge release on this machine
  join <endpoint>                  Join a network by providing a valid forge web graphql endpoint
  logs [type]                      Show logs for various forge components
  ls                               List forge releases installed locally
  ls:remote                        List remote forge releases available for install
  ps                               List running forge component processes
  remote [shellName]               Connects to the running system via a remote shell
  simulator [action]               Start/stop simulator and generate random traffic
  start [options] [<chainName>]    Start the forge and forge web deamon
  status [type]                    List info of the running chain/node
  stop [options] [<chainName>]     Stop the forge daemon and all related services
  tx [hash]                        Get a tx detail and display
  tx:ls                            List latest transactions
  upgrade [<chainName>]            Upgrade chain node to new version without reset
  use [version]                    Activate an already downloaded forge release
  version [<chainName>]            Output version for all forge components
  wallet:create                    Create a local wallet and dump its public/private key
  web [action]                     Open the web interface of running forge chain/node

Examples:

  Please install a forge-release before running any other commands
  > forge install latest
  > forge install --mirror https://releases.arcblockio.cn

  Curious about how to use a subcommand?
  > forge help install
  

```

## FAQ

Checkout [FAQ.md](./docs/FAQ.md)


CHANGELOG.md
## 1.3.0 (June 19, 2020)

- chore: bump forge sdk version

## 1.2.2 (April 22, 2020)

- chore: only consume starter blocklets from forge-cli
- chore: bump forge sdk

## 1.2.1 (March 26, 2020)

- use npm package's shasum to check
- fix forge web router bug
- add public network list

## 1.2.0 (March 26, 2020)

- chore: cleanup transaction protocol related code

## 1.1.3 (March 25, 2020)

- 添加跳过 npm 包校验命令行参数: `--no-verify`
- 设置 blocklet registry 为配置项

## 1.1.2 (March 21, 2020)

- modify descriptions after chain started

## 1.0.1 (March 20, 2020)

- retire dapp workshop
- retire forge swap
- bump version

## 1.1.0 (March 18, 2020)

- [one-webapp] serve static webapp via pm2 && serve-static

## 1.0.15 (March 09, 2020)

- fix blocklet:use bug

## 1.0.14 (March 04, 2020)

- add npm package integrity verification when run blocklet:use
- modify blocklet:use prompts

## 1.0.13 (February 29, 2020)

- optimizing install blocket dependencies

## 1.0.12 (December 15, 2019)

- chore: bump forge sdk version
- feat: support validator update in silent mode
- feat: support validator declare option
- feat: basic validator update features #354
- feat: support config the restricted declare feature in latest forge web #355

## 1.0.11 (December 04, 2019)

- hotfix: warning un supported OS error

## 1.0.10 (December 04, 2019)

- chore: remove useless app logs
- [unit test] libs.getForgeDistributionByOS
- feat: print more accurate OS distribution when install/download #348

## 1.0.9 (December 03, 2019)

- chore: bump version

## 1.0.8 (November 26, 2019)

- feat: retry download if failed #332

## 1.0.7 (November 26, 2019)

- fix unit test in makefile
- fix import module error
- remove unused empty line
- add test to script hooks
- feat: load remote forge releases when upgrade #334

## 1.0.6 (November 25, 2019)

- fix download patch version forge error

## 1.0.5 (November 25, 2019)

- refact: fetch latest releases
- refact: download module

## 1.0.4 (November 22, 2019)

- chore: modify 'forge web' to 'Forge Web'
- feat: adapt to latest forge swap configuration #341
- fix: compile contract error #342

## 1.0.3 (November 21, 2019)

- feat: print common tools version info if there was no chain name specified #333

## 1.0.2 (November 15, 2019)

- fix: CLI upgrade description typo

## 1.0.1 (November 13, 2019)

- chore: update cli doc links

## 1.0.0 (November 12, 2019)

- bump Forge CLI version to 1.0.0
- feat: upgrade forge js sdk to 1.0.0
- chore: not support strict upgrade rule temporarily
- chore: update forge quick start video in README
- chore: remove unnecessary code about chain from use command
- fix: forge releases sort bug when upgrade
- chore: rename prepare to deploy:prepare #300

## 0.40.7 (November 10, 2019)

- remove checkin function from create chain
- adapt latest workshop config

## 0.40.6 (November 09, 2019)

- fix: download prerelease version forge assets bug

## 0.40.5 (November 07, 2019)

- feat: use update-notifier package to check update
- chore: remove postinstall script && set min block time to 3s
- chore: remove moderator setting when config chain
- chore: remove ipfs code
- chore: modify moderator sk outputs when creat chain
- feat: modify the flow of using cli for the first time
- chore: modify typo in forge web and forge chain:config

## 0.40.4 (November 06, 2019)

- integrate atomic service

## 0.40.3 (November 04, 2019)

- chore: check if forge process exists before start web when run `forge start`
- improvement: validate ugprade tx during upgrade

## 0.40.2 (November 02, 2019)

- fix: create chain failed

## 0.40.1 (November 01, 2019)

- feat: modify the method of naming moniker
- improvement: forge version ui and upgrade limitation

## 0.40.0 (October 31, 2019)

- chore: remove create:cli code
- feat: upgrade js-sdk and deprecate commands

## 0.39.28 (October 31, 2019)

- fix: remove `app` from forge config

## 0.39.27 (October 30, 2019)

- breaking change: change account/blocklet/chain/contract/install/tx subcommands #300
- chore: remove non-root doc if run with root user #279
- feat: ignore fetch remote releases error when list local releases #299
- chore: bunch of updates
- refact: remove useless deps

## 0.39.26 (October 30, 2019)

- fix: cache failed

## 0.39.25 (October 29, 2019)

- feat: add contract:create command #164
- refact: rename protocol:ls to contract:ls
- refact: rename protocol:deploy to contract:deploy
- refact: rename protocol:deactivate to contract:deactivate
- refact: rename protocol:compile to contract:compile
- refact: rename protocol:activate to contract:activate

## 0.39.24 (October 28, 2019)

- refactor: rename config to chain:config
- feature: add forge config command #213
- chore: modify mechanism of check for latest

## 0.39.23 (October 25, 2019)

- fix: bug with multi user at same host #284

## 0.39.22 (October 25, 2019)

- fix: moderator is required by forge #291
- refact: move default icon to constant

## 0.39.21 (October 24, 2019)

- improvement: chain name is no longer required by forge version #290

## 0.39.20 (October 24, 2019)

- fix: fetch public IPv4 failed during forge prepare #305

## 0.39.19 (October 23, 2019)

- fix: CLI still fetch remote assets while `releaseDir` was set #302
- fix: `host` is require by latest workshop config #304

## 0.39.18 (October 23, 2019)

- validate if the chain config folder is complete #282

## 0.39.17 (October 22, 2019)

- add `cwd` param when reading ~/.forgerc.yml, to ensure read from current user's space #293;
- remove releases, rpc and chainName requirements from `forge ps` command.

## 0.39.16 (October 22, 2019)

- fix: start workshop failed with latest version #296

## 0.39.15 (October 18, 2019)

- chore: bump forge sdk version

## 0.39.14 (October 15, 2019)

- blocklet:init: add -d/-y args support
- feat: add blocklet:init command

## 0.39.13 (October 12, 2019)

- chore: update typo
- chore: make output more friendly if the protocol has been deployed
- fix: forge chain:ls not list all protocols

src/core/env.js
/* eslint-disable no-console */

const fs = require('fs');
const path = require('path');
const chalk = require('chalk');
const shell = require('shelljs');
const semver = require('semver');
const inquirer = require('inquirer');
const isElevated = require('is-elevated');
const { get, set, pickBy } = require('lodash');
const GRpcClient = require('@arcblock/grpc-client');
const { parse } = require('@arcblock/forge-config');

const {
  getChainDirectory,
  isChainExists,
  isDirectory,
  isFile,
  getChainReleaseFilePath,
  getChainWebConfigPath,
  getChainNameFromForgeConfig,
} = require('core/forge-fs');
const { ensureForgeRelease } = require('core/forge-config');
const { isForgeStarted, getProcessTag, getAllProcesses } = require('./forge-process');
const { printError, printInfo, printLogo, printWarning } = require('./util');
const { hasChains, getOSUserInfo, getTopChainName, cache } = require('./libs/common');

const { DEFAULT_CHAIN_NAME_RETURN, REQUIRED_DIRS } = require('../constant');
const { symbols, hr, pretty } = require('./ui');
const debug = require('./debug')('env');

const CURRENT_WORKING_CHAIN = getChainDirectory(process.env.FORGE_CURRENT_CHAIN);
process.env.CURRENT_WORKING_CHAIN = CURRENT_WORKING_CHAIN;

const config = { cli: {} }; // global shared forge-cli run time config

/**
 * Setup running env for various commands, the check order for each requirement is important
 * Since a running node requires a release directory
 * An unlocked wallet requires a valid config
 * @param {*} requirements
 * @param {boolean|function} requirements.runningNode Indicate whether the command need a running node
 * @param {boolean|string|function} requirements.chainName Indicate whether the command need `chainName` arg, possible: boolean, 'required', function
 * @param {boolean} requirements.chainExists Indicate whether the command need the `chain` to exists, if the attribute is true, the `requirements.chainName` will be true
 * @param {boolean} requirements.currentChainRunning Indicate whether the command need the `chain` to be running
 * @param {boolean|function} requirements.forgeRelease Indicate whether the command need forge release to exits
 * @param {boolean|function} requirements.wallet Indicate whether the command need a wallet
 * @param {boolean|function} requirements.rpcClient Indicate whether the command need RPC source
 * @param {*} args
 */
async function setupEnv(requirements, args = {}) {
  ensureEnv(args);
  await ensureNonRoot();
  ensureRequiredDirs();

  await ensureChainName(requirements.chainName, args);
  if (process.env.FORGE_CURRENT_CHAIN) {
    printInfo(`Working on ${chalk.cyan(process.env.FORGE_CURRENT_CHAIN)} chain`);
  }

  await ensureChainExists({
    requirement: requirements.chainExists,
    chainName: process.env.FORGE_CURRENT_CHAIN,
    configPath: process.env.FORGE_CONFIG,
  });

  await ensureCurrentChainRunning(
    requirements.currentChainRunning,
    process.env.FORGE_CURRENT_CHAIN
  );
  await ensureRunningChain(requirements.runningNode, process.env.FORGE_CURRENT_CHAIN);

  // Support evaluating requirements at runtime
  Object.keys(requirements).forEach(x => {
    if (typeof requirements[x] === 'function') {
      requirements[x] = requirements[x](args);
    }
  });

  if (requirements.forgeRelease || requirements.runningNode) {
    const cliConfig = await ensureForgeRelease({
      exitOn404: true,
      chainName: args.chainName,
      allowMultiChain: args.allowMultiChain,
    });
    Object.assign(config.cli, cliConfig);
  }

  ensureSetupScript(args);

  if (requirements.wallet || requirements.rpcClient || requirements.runningNode) {
    await ensureRpcClient(args, process.env.FORGE_CURRENT_CHAIN);
  }

  if (requirements.wallet) {
    await ensureWallet(args);
  }

  const result = {
    chainName: process.env.FORGE_CURRENT_CHAIN,
  };

  return pickBy(result, v => v !== undefined);
}

async function ensureRunningChain(requirement) {
  if (!requirement) {
    return;
  }

  const allProcesses = await getAllProcesses();
  if (allProcesses.length === 0) {
    printWarning('No running chains.');
    printInfo(`You can create a chain by ${chalk.cyan('forge chain:create')}, and start it.`);
    process.exit(0);
  }
}

async function ensureChainName(requirement = true, args) {
  if (args.chainName) {
    process.env.FORGE_CURRENT_CHAIN = args.chainName;
    return;
  }

  if (process.env.FORGE_CONFIG) {
    const chainId = getChainNameFromForgeConfig(process.env.FORGE_CONFIG);
    if (!chainId) {
      throw new Error(
        `Invalid config path: app name is invalid, config path: ${process.env.FORGE_CONFIG}`
      );
    }

    process.env.FORGE_CURRENT_CHAIN = chainId;
    return;
  }

  if (requirement) {
    if (typeof requirement === 'function') {
      const chainName = await requirement(args);
      if (chainName === DEFAULT_CHAIN_NAME_RETURN.NO_CHAINS) {
        printWarning(
          `There are no chains, please create it by run ${chalk.cyan('forge chain:create')}`
        );
        process.exit(0);
      } else if (chainName === DEFAULT_CHAIN_NAME_RETURN.NO_RUNNING_CHAINS) {
        printWarning('There are no running chains');
        process.exit(0);
      }

      process.env.FORGE_CURRENT_CHAIN = chainName;
    } else if (requirement === 'required') {
      printError(`${chalk.cyan('chainName')} argument is required`);
      process.exit(1);
    } else {
      const chainName = getTopChainName();
      if (!chainName) {
        printWarning(
          `There are no chains, please create it by run ${chalk.cyan('forge chain:create')}`
        );
        process.exit(0);
      }

      process.env.FORGE_CURRENT_CHAIN = chainName; // eslint-disable-line
    }
  }
}

async function ensureChainExists({ requirement = true, chainName, configPath }) {
  if (configPath) {
    const chainId = getChainNameFromForgeConfig(process.env.FORGE_CONFIG);

    if (chainId === chainName) {
      // if the chain name is specified by config path, no longer check wether the chain does exit
      return;
    }
  }

  if (chainName && requirement === true) {
    if (!hasChains()) {
      printError(
        `There are no chains found, please run ${chalk.cyan('forge chain:create')} to create.`
      );
      process.exit(1);
    }

    if (!isChainExists(chainName)) {
      printError(`Chain ${chainName} does not exist`);
      printInfo(`You can create it by run ${chalk.cyan(`forge chain:create ${chainName}`)}`);
      process.exit(1);
    }
  }
}

async function ensureCurrentChainRunning(requirement, chainName) {
  if (!requirement) {
    return true;
  }

  if (!(await isForgeStarted(chainName))) {
    printWarning(`Chain ${chalk.yellow(chainName)} is not started yet!`);
    printInfo(`Please run ${chalk.cyan(`forge start ${chainName}`)} first!`);
    process.exit(0);
  }

  return true;
}

/**
 * Ensure we have an global rpc client <grpc-client instance> before command run
 * Configure file priority: cli > env > release bundled
 *
 * @param {*} args
 */
function ensureRpcClient(args, chainName) {
  const socketGrpc = args.socketGrpc || process.env.FORGE_SOCK_GRPC;
  const configPath = getChainReleaseFilePath(chainName);

  if (socketGrpc) {
    printInfo(
      `${chalk.yellow(`Using custom grpc socket endpoint: ${process.env.FORGE_SOCK_GRPC}`)}`
    );
    const forgeConfig = {
      forge: {
        sockGrpc: socketGrpc,
        unlockTtl: 300,
        web: {
          port: 8210,
        },
      },
    };
    debug(`using forge-cli with remote node ${socketGrpc}`);
    Object.assign(config, forgeConfig);
  } else if (fs.existsSync(configPath)) {
    if (
      process.env.FORGE_CONFIG &&
      chainName === getChainNameFromForgeConfig(process.env.FORGE_CONFIG)
    ) {
      printInfo(`${chalk.yellow(`Using custom forge config: ${process.env.FORGE_CONFIG}`)}`);
    }

    const forgeConfig = parse(configPath);
    config.cli.forgeConfigPath = configPath;
    Object.assign(config, forgeConfig);
    debug(`${symbols.success} Using forge config: ${configPath}`);
    debug(`${symbols.success} Using forge config: ${pretty(config)}`);
  } else {
    throw new Error(
      `No valid rpc configuration, socketGrpc: ${socketGrpc}, configPath: ${configPath}`
    );
  }
}

/**
 * Ensure we have an unlocked wallet before run actual command { address, token }
 */
async function ensureWallet() {
  const wallet = cache.read('wallet');
  if (wallet && wallet.expireAt && wallet.expireAt > Date.now()) {
    debug(`${symbols.success} Use cached wallet ${wallet.address}`);
    config.cli.wallet = wallet;
    return;
  }

  debug(`${symbols.warning} no unlocked wallet found!`);

  const cachedAddress = wallet ? wallet.address : '';
  const questions = [
    {
      type: 'confirm',
      name: 'useCachedAddress',
      message: `Use cached wallet <${cachedAddress}>?`,
      when: !!cachedAddress,
    },
    {
      type: 'text',
      name: 'address',
      default: answers => (answers.useCachedAddress ? cachedAddress : ''),
      when: answers => !cachedAddress || !answers.useCachedAddress,
      message: 'Please enter wallet address:',
      validate: input => {
        if (!input) return 'The address should not empty';
        return true;
      },
    },
    {
      type: 'password',
      name: 'passphrase',
      message: 'Please enter passphrase of the wallet:',
      validate: input => {
        if (!input) return 'The passphrase should not empty';
        if (!/^.{6,15}$/.test(input)) return 'The passphrase must be 6~15 chars long';
        return true;
      },
    },
  ];

  const client = createRpcClient();
  const { address: userAddress, passphrase, useCachedAddress } = await inquirer.prompt(questions);
  const address = useCachedAddress ? cachedAddress : userAddress;
  try {
    const { token } = await client.loadWallet({ address, passphrase });
    cache.write('wallet', {
      address,
      token,
      expireAt: Date.now() + (config.forge.unlockTtl || 300) * 1e3,
    });
    config.cli.wallet = { address, token };
    debug(`${symbols.success} Use unlocked wallet ${address}`);
  } catch (err) {
    debug.error('Wallet Unlock Error', err);
    shell.echo(`${symbols.error} wallet unlock failed, please check wallet address and passphrase`);
    process.exit(1);
  }
}

/**
 * If we have application specific protobuf, we need to load that into sdk
 *
 * @param {*} args
 */
function ensureSetupScript(args) {
  const setupScript = args.setupScript || process.env.FORGE_SDK_SETUP_SCRIPT;
  if (setupScript && fs.existsSync(setupScript) && fs.statSync(setupScript).isFile()) {
    debug(`${symbols.warning} loading custom scripts: ${setupScript}`);
    // eslint-disable-next-line
    require(path.resolve(setupScript));
  }
}

/**
 * Find version of current forge release
 *
 * @param {*} releaseDir
 * @returns String
 */
function findReleaseVersion(releaseDir) {
  if (!releaseDir) {
    return '';
  }

  const parentDir = path.join(releaseDir, 'forge/releases');
  if (!isDirectory(parentDir)) {
    return '';
  }

  return fs
    .readdirSync(parentDir)
    .find(x => fs.statSync(path.join(parentDir, x)).isDirectory() && semver.valid(x));
}

async function ensureNonRoot() {
  try {
    if (await isElevated()) {
      printError(`${chalk.red('Error: forge cannot be started with root permission')}`);
      process.exit(77);
    }
  } catch (err) {
    debug.error(err);
    printError('Cannot get current username');
  }
}

function ensureEnv({ configPath }) {
  if (process.env.FORGE_CONFIG) {
    if (!fs.existsSync(process.env.FORGE_CONFIG)) {
      throw new Error(`env FORGE_CONFIG is invalid: ${process.env.FORGE_CONFIG}`);
    }
  } else if (configPath) {
    if (!fs.existsSync(configPath)) {
      throw new Error(`config path does not exit, config path: ${configPath}`);
    }

    process.env.FORGE_CONFIG = configPath;
  }
}

/**
 * Ensure we have required directories done
 */
function ensureRequiredDirs() {
  Object.keys(REQUIRED_DIRS).forEach(x => {
    const dir = REQUIRED_DIRS[x];
    if (isDirectory(dir)) {
      debug(`${symbols.info} ${x} dir already initialized: ${dir}`);
    } else {
      try {
        shell.mkdir('-p', dir);
        debug(`initialized ${x} dir for forge-cli: ${dir}`);
      } catch (err) {
        shell.echo(`${symbols.success} failed to initialize ${x} dir for forge-cli: ${dir}`, err);
      }
    }
  });
}

let client = null;
function createRpcClient() {
  if (client) {
    return client;
  }

  const sockGrpc =
    process.env.FORGE_SOCK_GRPC || get(config, 'forge.sockGrpc') || get(config, 'forge.sock_grpc');

  printInfo(`Connect to grpc endpoint: ${sockGrpc}`);
  client = new GRpcClient(sockGrpc);
  return client;
}

// TODO: need to refact to on name parameter
/**
 * @deprecated Use core/libs/common/makeNativeCommandRunnder
 * @param {*} executable
 * @param {*} name
 * @param {*} param2
 */
function makeNativeCommandRunner(executable, name, { env } = {}) {
  return function runNativeForgeCommand(subCommand, options = {}) {
    return function rumCommand() {
      // eslint-disable-next-line prefer-destructuring
      const forgeConfigPath = config.cli.forgeConfigPath;
      const binPath = config.cli[executable];

      if (!binPath) {
        shell.echo(`${symbols.error} ${executable} not found, abort!`);
        shell.echo(`${symbols.info} we recommend a clean setup from scratch with following steps:`);
        shell.echo(hr);
        shell.echo(`${chalk.cyan('forge stop')}`);
        shell.echo(`${chalk.cyan('rm -rf ~/.forge_cli')}`);
        shell.echo(`${chalk.cyan('rm -rf ~/.forge_chains')}`);
        shell.echo(`${chalk.cyan('npm install -g @arcblock/forge-cli')}`);
        shell.echo(`${chalk.cyan('forge install')}`);
        shell.echo('');
        return process.exit(1);
      }

      const sockGrpc =
        process.env.FORGE_SOCK_GRPC || get(config, 'forge.sockGrpc') || 'tcp://127.0.0.1:28210';

      const erlAflagsParam = `ERL_AFLAGS="-sname ${getProcessTag(
        name,
        process.env.FORGE_CURRENT_CHAIN
      )}"`;
      let command = `${erlAflagsParam} FORGE_CONFIG=${forgeConfigPath} ${binPath} ${subCommand}`;

      if (executable === 'webBinPath') {
        const webConfigPath = getChainWebConfigPath(name);
        if (isFile(webConfigPath)) {
          command = `${erlAflagsParam} FORGE_WEB_CONFIG=${webConfigPath} FORGE_SOCK_GRPC=${sockGrpc} ${binPath} ${subCommand}`; // eslint-disable-line
        } else {
          command = `${erlAflagsParam} FORGE_SOCK_GRPC=${sockGrpc} ${binPath} ${subCommand}`; // eslint-disable-line
        }
      }

      if (executable === 'simulatorBinPath') {
        command = `${erlAflagsParam} FORGE_SOCK_GRPC=${sockGrpc} ${binPath} ${subCommand}`; // eslint-disable-line
      }

      if (env) {
        debug('makeNativeCommandRunner env:', env);
        command = `${env} ${command}`;
      }

      const { shell: envShell, homedir } = getOSUserInfo();
      command = `SHELL=${envShell} HOME=${homedir} ${command}`;

      debug(`runNativeCommand.${executable}:`, command);
      return shell.exec(command, options);
    };
  };
}

// Because some comments have special usage, we need to add it back
function ensureConfigComment(str) {
  return str.replace(
    '[[tendermint.genesis.validators]]',
    '### begin validators\n[[tendermint.genesis.validators]]\n### end validators'
  );
}

debug.error = (...args) => {
  if (debug.enabled) {
    console.error(...args);
  }
};

module.exports = {
  config: {
    get: (key, defaultValue) => get(config, key, defaultValue),
    set: (key, value) => set(config, key, value),
  },

  webUrl() {
    return `http://localhost:${get(config, 'forge.web.port') || 8210}`;
  },

  debug,
  setupEnv,
  findReleaseVersion,
  ensureRequiredDirs,
  ensureRpcClient,
  makeNativeCommandRunner,
  runNativeWebCommand: makeNativeCommandRunner('webBinPath', 'web'),
  runNativeSimulatorCommand: makeNativeCommandRunner('simulatorBinPath', 'simulator'),
  createRpcClient,
  isDirectory,
  printLogo,
  ensureConfigComment,
};
